ast_tools/passes/cse.py

```diff
@@ -0,0 +1,197 @@
+import ast
+import astor
+from copy import deepcopy
+from collections import Counter
+import typing as tp
+import warnings
+
+from . import Pass
+from . import PASS_ARGS_T
+
+from ast_tools.common import gen_free_prefix, gen_free_name, is_free_name
+from ast_tools.immutable_ast import immutable, mutable
+from ast_tools.stack import SymbolTable
+from ast_tools.transformers.node_replacer import NodeReplacer
+from ast_tools.visitors.node_finder import NodeFinder
+
+__all__ = ['cse']
+
+def _is_leaf_expr(node: ast.expr):
+    assert isinstance(nod, ast.expr)
+
+    return isinstance(node,(
+        ast.Attribute,
+        ast.Constant,
+        ast.Name,
+        ast.NameConstant,
+        ast.Num,
+        ast.Subscript,
+        ))
+
+class ExprFinder(NodeFinder):
+    def _get_key(self, node):
+        if isinstance(node, ast.expr):
+            # Need the immutable value so its comparable
+            return immutable(node)
+        else:
+            return None
+
+class ExprReplacer(NodeReplacer):
+    def _get_key(self, node):
+        if isinstance(node, ast.expr):
+            # Need the immutable value so its comparable
+            return immutable(node)
+        else:
+            return None
+
+class ExprCounter(ast.NodeVisitor):
+    def __init__(self):
+        self.cses = Counter()
+
+    def visit_UnaryOp(self, node: ast.UnaryOp):
+        self.cses[immutable(node)] += 1
+        self.generic_visit(node)
+
+    def visit_BinOp(self, node: ast.BinOp):
+        self.cses[immutable(node)] += 1
+        self.generic_visit(node)
+
+    def visit_BoolOp(self, node: ast.BoolOp):
+        self.cses[immutable(node)] += 1
+        self.generic_visit(node)
+
+    def visit_Compare(self, node: ast.Compare):
+        self.cses[immutable(node)] += 1
+        self.generic_visit(node)
+
+    def visit_IfExpr(self, node: ast.IfExp):
+        self.cses[immutable(node)] += 1
+        self.generic_visit(node)
+
+    def vist_Call(self, node: ast.Call):
+        # this seems hard to get right so not going to save Calls
+        return self.generic_visit(node)
```

Comment:
I'm guessing one problem with Calls is side effects.  Should we be concerned with any of the above expressions containing side effect code? E.g. we could have a expression that contains a call that is contained within another expression that is commonly shared.

Reply:
Given that this depends on SSA, I'm guessing that we'll just be restricting this to side-effect free code for now?

Reply:
Yeah I was worried about side effects.   So I suppose we could save calls.

Reply:
I will make it optional